File: packages/platform-go/observability/observability.go
// Package observability provides the OpenTelemetry tracer + meter
// provider initialisation shared by every Go service. Mirrors the
// `platform-spring-boot-observability` Kotlin module and the
// `platform_python.observability` package. The MDC field "request_id"
// is the canonical correlation hook (ADR-0019).
package observability

import (
	"context"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

// Config holds the OTel config knobs.
type Config struct {
	Service    string
	Env        string
	Region     string
	Tenant     string
	OTLPEndpoint string
	UseStdoutFallback bool
}

// DefaultConfig returns the platform defaults.
func DefaultConfig(service string) Config {
	return Config{
		Service:  service,
		Env:      os.Getenv("PLATFORM_ENV"),
		Region:   os.Getenv("PLATFORM_REGION"),
		Tenant:   os.Getenv("PLATFORM_TENANT"),
		OTLPEndpoint: os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT"),
	}
}

// ShutdownFunc is returned by InitTracer and must be called when
// the service stops.
type ShutdownFunc func(context.Context) error

// InitTracer initialises the global TracerProvider. Returns a shutdown
// function that must be called on service exit. Falls back to the
// stdout exporter when OTLP endpoint is not set.
func InitTracer(ctx context.Context, cfg Config) (ShutdownFunc, error) {
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(cfg.Service),
			semconv.DeploymentEnvironment(cfg.Env),
			semconv.CloudRegion(cfg.Region),
		),
	)
	if err != nil {
		return nil, err
	}

	var exporter sdktrace.SpanExporter
	if cfg.OTLPEndpoint != "" {
		exporter, err = otlptrace.New(ctx, otlptracegrpc.NewClient(
			otlptracegrpc.WithEndpoint(cfg.OTLPEndpoint),
			otlptracegrpc.WithInsecure(),
		))
		if err != nil {
			return nil, err
		}
	} else if cfg.UseStdoutFallback {
		exporter, err = stdouttrace.New(stdouttrace.WithPrettyPrint())
		if err != nil {
			return nil, err
		}
	} else {
		// No exporter — use a no-op.
		otel.SetTracerProvider(sdktrace.NewTracerProvider(
			sdktrace.WithSampler(sdktrace.NeverSample()),
		))
		return func(c context.Context) error { return nil }, nil
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter, sdktrace.WithBatchTimeout(5*time.Second)),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))
	return tp.Shutdown, nil
}

File: packages/platform-python/src/platform_python/observability.py
"""OpenTelemetry tracer + meter provider initialisation shared by every Python service.

Mirrors ``platform-spring-boot-observability`` (Kotlin) and
``platform-go/observability`` (Go).
"""
from __future__ import annotations

import logging
import os
from typing import Any

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
from opentelemetry.semconv.resource import ResourceAttributes

log = logging.getLogger(__name__)

_initialized = False


def init_tracer(
    service_name: str,
    *,
    env: str | None = None,
    region: str | None = None,
    tenant: str | None = None,
    otlp_endpoint: str | None = None,
    console_fallback: bool = False,
) -> trace.Tracer:
    """Initialise the global TracerProvider. Idempotent.

    Args:
        service_name: Human-readable service name (e.g. ``fraud-risk-service``).
        env: Environment tag (``dev``, ``stg``, ``prod``). Defaults to ``PLATFORM_ENV``.
        region: Region tag (e.g. ``eu-west-1``). Defaults to ``PLATFORM_REGION``.
        tenant: Tenant slug. Defaults to ``PLATFORM_TENANT``.
        otlp_endpoint: OTLP gRPC endpoint. Defaults to ``OTEL_EXPORTER_OTLP_ENDPOINT``.
        console_fallback: If True and no OTLP endpoint, writes spans to stdout.
    """
    global _initialized
    if _initialized:
        return trace.get_tracer(service_name)

    env = env or os.getenv("PLATFORM_ENV", "dev")
    region = region or os.getenv("PLATFORM_REGION", "local")
    tenant = tenant or os.getenv("PLATFORM_TENANT", "global")
    otlp_endpoint = otlp_endpoint or os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "")

    resource = Resource.create(
        {
            ResourceAttributes.SERVICE_NAME: service_name,
            ResourceAttributes.DEPLOYMENT_ENVIRONMENT: env,
            "region": region,
            "tenant": tenant,
        }
    )
    provider = TracerProvider(resource=resource)
    if otlp_endpoint:
        provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=otlp_endpoint, insecure=True)))
    elif console_fallback:
        provider.add_span_processor(BatchSpanProcessor(ConsoleSpanExporter()))
    trace.set_tracer_provider(provider)
    _initialized = True
    log.info("platform-python.tracer.init service=%s env=%s region=%s otlp=%s", service_name, env, region, otlp_endpoint)
    return trace.get_tracer(service_name)


def get_tracer(name: str) -> trace.Tracer:
    """Return the named tracer (initialises the provider with default config if not yet set)."""
    if not _initialized:
        init_tracer(name)
    return trace.get_tracer(name)
